### DISCUSSION TITLE
How to make hostid the same as underlying machine?

### DISCUSSION THREAD

**Discussion by venaturum** [netblue30/firejail#6375] [Q&A]
Apologies as I know very little about firejail and do not use it, however I am trying to resolve an issue on behalf of a customer using our software.  Licenses issued to users are based on the hostid of the machine.  Firejail under default settings sets the hostid to 0 and consequently the license can't be used.  Is there a workaround?

**Comment by ghost** [comment]
Hi. Firejail uses application-specific profiles for sandboxing. Basically these are 'recipes' used to create sandboxes that usually get installed under `/etc/firejail` by the package management tooling provided by the OS. Users can place overrides to these system-wide profiles in their `${HOME}/.config/firejail`. In general, when Firejail can't find a profile that matches the executable name of a program asked to run sandboxed, it uses a default set of rules to create one.

I did a quick check on Arch Linux and a Ubuntu 22.04 LTS machine and see the same hostid in- and outside the firejail sandbox.

```sh
$ hostid
007f0100
```

```sh
$ firejail /usr/bin/hostid
007f0100
```

Do you happen to know how your customer starts this program and what the program's executable is called? It might be needing specific access to something under /etc or wherever the license details of your software is placed in the filesystem. Without this information it's guessing what's going on exactly and to provide possible workarounds/fixes.

HTH

**Reply by venaturum** [reply]
Thanks @glitsj16, their machine is linux and the HostID is taken from the last eight hexadecimal digits of the MAC address of the primary network adapter.

There is several ways users can invoke our program.  One is via a command line executable `gurobi_cl arg1 arg2 argn` or they can also call it from a various languages via APIs we provide, eg `python my_script_containing_gurobi_calls.py`.

**Reply by ghost** [reply]
Aha, that's important information, thanks. Firejail sure can (and does) block `python` access in a lot of its profiles. Typically this can be allowed by adding `--include=allow-python3.inc` on CLI or `include allow-python3.inc` in a file placed under `~/.config/firejail/foo.local` that will get parsed automatically when the program's executable is called `foo`. That's something you can advise your customer to try so things work as expected from inside the sandbox.

**Reply by venaturum** [reply]
That could be very helpful, but I'm not sure they're being blocked from running our program - they do see output from our program, which includes an error message `HostID mismatch (licensed to ae4de676, hostid is 0)`.  So it seems to be a mac address issue, or perhaps there's something about firejail which makes our code to read the mac address fail.

**Reply by ghost** [reply]
If they can run python as you indicate, my above mitigation remark is irrelevant. I can see why you think it's probably a mac address issue, agreed. Although Firejail supports assigning a (random) mac address, this is actually an option that is not included in any of our sandboxing profiles by default. In other words, the user should have had to enable that explicitly, and be fully aware of doing so. I'm assuming there's something else in play that I'm unable to pinpoint without knowing how exactly they are starting your software in a firejail sandbox. A few things to consider: (1) ask them for a debug log via firejail's `--debug` flag, and (2) ask them to run a `strace` so you can help determine what files (under /etc probably) are being accessed.

**Reply by venaturum** [reply]
Ok will do, thank you very much for the help!

**Comment by kmk3** [comment]
Basic debugging information is missing, so it's hard to tell where exactly the
problem is.

> Apologies as I know very little about firejail and do not use it, however I
> am trying to resolve an issue on behalf of a customer using our software.

No problem; can you ask the user to fill the bug report template or to just
open a bug report directly?

Link:

* <https://github.com/netblue30/firejail/issues/new?template=bug_report.md>

**Comment by kmk3** [comment]
(Offtopic)

> their machine is linux and the HostID is taken from the last eight
> hexadecimal digits of the MAC address of the primary network adapter.

Note that /etc/hostid is not guaranteed to exist, in which case apparently the
glibc version of `hostid` is supposed to generate something based on the
hostname and/or the local IPv4 address, neither of which are likely to be
unique.

From `gethostname(3)`:

> In the glibc implementation, the hostid is stored in the file /etc/hostid.
> (Before glibc 2.2, the file /var/adm/hostid was used.)
> 
> In the glibc implementation, if gethostid() cannot open the file containing
> the host ID, then it obtains the hostname using gethostname(2), passes that
> hostname to gethostbyname_r(3) in order to obtain the host's IPv4 address,
> and returns a value obtained by bit- twiddling the IPv4 address.  (This value
> may not be unique.)

Also, hardware can be changed/upgraded and hardware IDs may not be unique
inside VMs, so it seems like it would be more reliable to use an
installation-specific ID.

The most common one on Linux is probably the `machine-id` file, which contains
a large hexadecimal number and is usually generated by dbus.

It is usually in one of the following paths:

* /etc/machine-id
* /var/lib/dbus/machine-id

I'd try to use those first and only use `hostid` as a fallback.

See also `dbus-uuidgen(1)`.

**Comment by rusty-snake** [comment]
Tested with allmost all firejail options we use in profiles and hostid still works. So something else must be going on.

Also knowing the profile that is used could help. If it is our `default.profile` I would guess `disable-common.inc` or `protocol` (missing netlink). Otherwise it might be worth to check that `firejail --noprofile <program and args>` works.
